python/pyarrow/_flight.pyx

```diff
@@ -874,9 +891,38 @@ cdef class FlightClient:
 
         with nogil:
             check_flight_status(CFlightClient.Connect(c_location, c_options,
-                                                      &result.client))
+                                                      &self.client))
 
-        return result
+    def wait_for_available(self, timeout=5):
+        """Block until the server can be contacted.
+
+        Parameters
+        ----------
+        timeout : int, default 5
+            The maximum seconds to wait.
+        """
+        deadline = time.time() + timeout
+        while True:
+            try:
+                list(self.list_flights())
```

Comment:
@pitrou additionally this method uses `list_flights` which raises if not implemented on the server. The string containing check is pretty error prone below, so we would require a ping call instead. Jira?

Reply:
You should be able to only consider [`FlightUnavailableError`](https://github.com/apache/arrow/blob/master/python/pyarrow/_flight.pyx#L138) (and perhaps `FlightTimedOutError`) as a failure.

Reply:
Thanks, updated!